hazelcast/src/main/java/com/hazelcast/internal/affinity/ThreadAffinitySupport.java

```diff
@@ -0,0 +1,164 @@
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.hazelcast.internal.affinity.ThreadAffinity.Group.IO;
+import static com.hazelcast.internal.affinity.ThreadAffinity.Group.PARTITION_THREAD;
+import static com.hazelcast.internal.affinity.ThreadAffinityProperties.isAffinityEnabled;
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent;
+import static com.hazelcast.util.EmptyStatement.ignore;
+import static com.hazelcast.util.OsHelper.OS;
+import static com.hazelcast.util.OsHelper.isUnixFamily;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.util.Collections.disjoint;
+
+public class ThreadAffinitySupport {
+
+    public static volatile ThreadAffinitySupport INSTANCE = null;
+    private static final AtomicBoolean INIT = new AtomicBoolean(false);
+
+    private final ILogger logger;
+    private ConcurrentMap<ThreadAffinity, ThreadAffinityController> registry = new ConcurrentHashMap<>();
+
+    private ThreadAffinitySupport(ILogger logger) {
+        this.logger = logger;
+    }
+
+    public ThreadAffinityController create(Class<? extends Thread> cls) {
+        checkNotNull(cls);
+        ThreadAffinity type = cls.getAnnotation(ThreadAffinity.class);
+        if (type != null) {
+            return getOrPutIfAbsent(registry, type, key -> new ThreadAffinityController(logger, type));
+        }
+
+        return null;
+    }
+
+    public static void init(ILogger logger) {
+        if (!isAffinityEnabled()) {
+            return;
+        }
+
+        if (!INIT.compareAndSet(false, true)) {
+            throw new IllegalStateException("Thread affinity support disallows multiple Hazelcast instances under the same JVM");
+        }
+
+        if (!isUnixFamily()) {
+            throw new IllegalStateException("Thread affinity not supported on this platform: " + OS);
+        }
+
+        logger.info("Thread affinity option detected, checking for dependencies");
+
+        try {
+            // Check if dependency OpenHFT Affinity is present in classpath
+            Class.forName("net.openhft.affinity.AffinityLock");
+        } catch (ClassNotFoundException e) {
+            logger.warning("Dependency net.openhft:affinity not found in classpath. Affinity support is disabled.");
+            return;
+        }
+
+        try {
+            // Check if JNA is available in the classpath
+            Class.forName("com.sun.jna.Platform");
+        } catch (ClassNotFoundException e) {
+            logger.warning("Dependency net.java.dev.jna:jna* not found in classpath. Affinity support is disabled.");
+        }
+
+        if (!disjoint(ThreadAffinityProperties.getCoreIds(IO), ThreadAffinityProperties.getCoreIds(PARTITION_THREAD))) {
+            logger.warning("Affinity assignments for the different affinity groups have some cores in common (shared).");
```

Comment:
Perhaps best to throw an exception so that it is very obvious the system isn't working as specified.